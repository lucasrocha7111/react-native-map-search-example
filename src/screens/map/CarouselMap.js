import React from 'react'
import {
    View,
    Text,
    Platform
} from 'react-native'
import Styles, { Colors, width } from '../../themes/default/styles'
import Carousel from 'react-native-snap-carousel'

export class CarouselMap extends React.Component {

    render() {
        return (
            <Carousel
                ref={(c) => { this._carousel = c; }}
                data={this.props.entries}
                renderItem={this._renderItem}
                sliderWidth={width}
                itemWidth={300}
                style={[Styles.shadow, {height: 80, backgroundColor: 'red'}]}
                onSnapToItem={(index) => {
                    this.props.moveMap(this.props.entries[index].coordinate)
                }}
            />
        )
    }

    _renderItem ({item, index}) {
        return (
            <View key={index} style={[Platform.OS === 'ios' ? Styles.shadow : null, {justifyContent: 'center', backgroundColor: Colors.white, paddingVertical: 20, height: 80, borderRadius: 2}]}>
                <Text style={[Styles.regularFont, {textAlign: 'center'}]}>{ item.title }</Text>
            </View>
        );
    }
}