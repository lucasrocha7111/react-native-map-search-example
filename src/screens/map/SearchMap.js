import React from 'react'
import {
    View,
    TextInput,
    TouchableOpacity,
    Text,
    ScrollView,
    Keyboard
} from 'react-native'
import Styles, { Colors } from '../../themes/default/styles'
import { GoogleMapsUtils } from '../../utils/google-maps'

export class SearchMap extends React.Component {

    state = {
        search: '',
        predictionList: []
    }

    mapUtils = new GoogleMapsUtils()
    timeout = null

    render() {
        return (
            <View style={[{width: '100%', paddingHorizontal: 30, top: 70, zIndex: 2, position: 'absolute'}]}>
                <View>
                    <TextInput 
                        style={[Styles.searchnput]} 
                        placeholder={'Digite uma localização'} 
                        value={this.state.search}
                        onChangeText={(value) => {
                            console.log('on change ', value)
                            this.setState({
                                search: value
                            })
                            this._onSearch(value)
                        }}
                        clearTextOnFocus={true}
                    />
                </View>
                <View style={[Styles.center, this.state.predictionList.length ? {minHeight: 300, width: '100%', marginHorizontal: 4, paddingHorizontal: 10, backgroundColor: Colors.white, borderRadius: 10, top: -10} : {zIndex: -1}]}>
                    <ScrollView keyboardShouldPersistTaps={'always'}>
                        {this.state.predictionList.map((data, index) => {
                            return (
                                <TouchableOpacity key={index} style={[{marginVertical: 10}]} onPress={() => {
                                    this._selectItem(data.text)
                                }}>
                                    <Text style={[Styles.regularFont]}>{data.text}</Text>
                                </TouchableOpacity>
                            )
                        })}
                    </ScrollView>
                </View>
            </View>
        )
    }

    _selectItem = async (item) => {
        Keyboard.dismiss()
        this.setState({
            search: item,
            predictionList: []
        })
        let res = await this.mapUtils.getLatLngByAddress(item)
        if(res.data.status === 'OK') {
            let location = {
                latitude: res.data.results[0].geometry.location.lat,
                longitude: res.data.results[0].geometry.location.lng,
            }
            this.props.onSelectAddress(location)
        }
    }

    _onSearch = (value) => {
        if(this.timeout !== null) {
            clearInterval(this.timeout)
        }
        this.timeout = setTimeout(async () => {
            let res = await this.mapUtils.getPlacesBySearch(value)
            if(res.data.status === 'OK') {
                let list = res.data.predictions
                let predictionList = list.map((data) => {
                    return {
                        text: data.description
                    }
                })
                this.setState({
                    predictionList: predictionList
                })
            }
        }, 800)
    }

}