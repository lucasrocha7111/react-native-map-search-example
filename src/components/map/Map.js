import React from 'react'
import { View } from 'react-native'
import MapView, { Marker } from 'react-native-maps'
import { Alert } from 'react-native'
import * as Location from 'expo-location'
import * as Permissions from 'expo-permissions'
import { CarouselMap } from './CarouselMap'
import { GoogleMapsUtils } from '../../utils/google-maps'

export const LATITUDE_DELTA = 0.0010
export const LONGITUDE_DELTA = 0.0020

export class Map extends React.Component {

    state = {
        location: null,
        markers: [],
        requestingPermisson: false
    }

    map = null
    mapUtils = new GoogleMapsUtils()

    render() {
        return (
            <View style={[{flex: 1}]}>
                <MapView 
                    ref={(c) => { this.map = c }}
                    style={[{flex: 1}]} 
                    showsUserLocation={true}
                    onMapReady={async () => {
                        await this._getLocationAsync()
                        setTimeout(() => {
                            this.getUserLocation()
                        }, 100)
                    }}
                >
                    {this.state.markers.map((data, index) => {
                        return (
                            <Marker key={index} title={data.title} coordinate={data.coordinate} />
                        )
                    })}
                </MapView>
                {this.state.markers.length
                ? <View style={[{position: 'absolute', bottom: 0, height: 100, justifyContent: 'center', paddingTop: 20 }]}>
                    <CarouselMap 
                        entries={this.state.markers} 
                        moveMap={(region) => {
                            let regionParams = {
                                latitude: region.latitude,
                                longitude: region.longitude,
                                latitudeDelta: LATITUDE_DELTA,
                                longitudeDelta: LONGITUDE_DELTA
                            }
                            this.map.animateToRegion(regionParams, 700)
                        }}
                    />
                </View>
                : null
                }
            </View>
        )
    }

    _getLocationAsync = async () => {
        if(!this.state.requestingPermisson) {
            let { status } = await Permissions.askAsync(Permissions.LOCATION)
            this.setState({
                requestingPermisson: true
            })
            if (status !== 'granted') {
                this.setState({
                    requestingPermisson: false
                })
                setTimeout(() => {
                    Alert.alert('Localização', 'Precisamos da sua localização', [
                        {text: 'Ok', onPress:() => {
                            this._getLocationAsync()
                        }}
                    ])
                }, 50)
            } else {
                let location = await Location.getCurrentPositionAsync({accuracy: Location.Accuracy.Balanced, maximumAge: 10000})
                this.setState({ location })
            }
        }
        
    }

    getUserLocation = () => {
        let { location } = this.state
        if(location !== null) {
            let region = {
                latitude: location.coords.latitude,
                longitude: location.coords.longitude,
                latitudeDelta: LATITUDE_DELTA,
                longitudeDelta: LONGITUDE_DELTA
            }
            this.map.animateToRegion(region, 1000)
            this.getNearbyPlaces(region)
        }
    }

    getNearbyPlaces = async (region) => {
        let nearby = await this.mapUtils.getNearbyPlaces(region)
        let markers = nearby.data.results.map((data) => {
            return {
                title: data.name,
                coordinate: {
                    latitude: data.geometry.location.lat,
                    longitude: data.geometry.location.lng,
                }
            }
        })
        this.setState({
            markers: markers
        })
        
    }
}