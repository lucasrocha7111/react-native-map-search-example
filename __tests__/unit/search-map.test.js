import React from 'react'
import renderer from 'react-test-renderer'
import { SearchMap } from '../../src/screens/map/SearchMap'


describe('<SearchMap />', () => {
    it('has 2 childrens', () => {
        const tree = renderer.create(<SearchMap />).toJSON()
        expect(tree.children.length).toBe(2)
    })
    it('renders correctly', () => {
        const tree = renderer.create(<SearchMap />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})