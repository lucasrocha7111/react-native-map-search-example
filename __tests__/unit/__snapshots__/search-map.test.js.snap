// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchMap /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "paddingHorizontal": 30,
        "position": "absolute",
        "top": 70,
        "width": "100%",
        "zIndex": 2,
      },
    ]
  }
>
  <View>
    <TextInput
      allowFontScaling={true}
      clearTextOnFocus={true}
      onChangeText={[Function]}
      placeholder="Digite uma localização"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#fff",
            "borderRadius": 20,
            "borderWidth": 0.7,
            "color": "#676e7d",
            "elevation": 3,
            "fontFamily": "lato",
            "fontSize": 14,
            "height": 42,
            "marginBottom": 20,
            "paddingHorizontal": 20,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 4,
            "width": "100%",
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "alignSelf": "center",
        },
        Object {
          "zIndex": -1,
        },
      ]
    }
  >
    <RCTScrollView
      keyboardShouldPersistTaps="always"
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
